### PR TITLE
make required fields configurable

### LINKED ISSUE #897: Feature Request: Pflichtfelder im Config definieren
Pflichtfelder als Confligflag und nicht hardcoded.

### PR DESCRIPTION
**Pull request by xuwhite**
I purposely not implemented planned_arrival_date in the config because at the moment it is not possible to not pick an arrival date even if it is not required. I think it makes no sense implementing it without messing with the date picker which is out of scope for me.
Fixes #897

### CODE REVIEW

File: src/Controllers/SignUpController.php [hunk 1]
```diff
@@ -101,6 +102,12 @@ private function renderSignUpPage(): Response
                 'isPronounEnabled' => $this->config->get('enable_pronoun'),
                 'isFullNameEnabled' => $this->config->get('enable_user_name'),
                 'isPlannedArrivalDateEnabled' => $this->config->get('enable_planned_arrival'),
+                'isPronounRequired' => $requiredFields['pronoun'],
```
Comment (MyIgel) [inline src/Controllers/SignUpController.php:105]: Using variables here imho does not make it better readable but as it done so before thats ok (and should definitely be refactored at a later stage)

File: src/Factories/User.php [hunk 2]
```diff
@@ -67,6 +67,16 @@ public function determineBuildUpStartDate(): DateTimeInterface
         return $this->config->get('buildup_start') ?? CarbonImmutable::now();
     }
 
+    private function required(string $key): string
+    {
+        $requiredFields = $this->config->get('signup_required_fields');
+        $isRequired = 'optional';
```
Comment (MyIgel) [inline src/Factories/User.php:73]:
```suggestion
        return $requiredFields[$key] ? 'required' : 'optional';
```

File: tests/Utils/SignUpConfig.php [hunk 3]
```diff
@@ -27,10 +34,18 @@ public static function setMaximumConfig(Config $config): void
         $config->set('enable_user_name', true);
         $config->set('enable_mobile_show', true);
         $config->set('enable_dect', true);
+        $config->set('signup_required_fields', $requiredFields);
     }
 
     public static function setMinimumConfig(Config $config): void
     {
+        $requiredFields = $config->get('signup_required_fields');
```
Comment (MyIgel) [inline tests/Utils/SignUpConfig.php:42]: (same here)

**Review by MyIgel — CHANGES_REQUESTED** [review]
I have some smaller suggestions to this ;)

File: config/config.default.php [hunk 4]
```diff
@@ -233,6 +233,16 @@
     // Users are able to sign up
     'registration_enabled'    => (bool) env('REGISTRATION_ENABLED', true),
 
+    // Required fields on sign-up page
+    'signup_required_fields'   => [
+        'pronoun'              => (bool) env('PRONOUN_REQUIRED', false),
```
Comment (MyIgel) [inline config/config.default.php:238]: Spacing might be improved too ;)

**Comment by MyIgel** [comment]
Another point is that
* a user currently can remove some data after sign up in their profile
* There is no hint implemented that notifies that some fields are missing when the config gets changed during an evnet

File: src/Controllers/SignUpController.php (hunk 1 above)
Comment (xuwhite) [reply src/Controllers/SignUpController.php:105]: I like it that way!
won't change

File: tests/Utils/SignUpConfig.php [hunk 5]
```diff
@@ -11,6 +11,13 @@ final class SignUpConfig
 {
     public static function setMaximumConfig(Config $config): void
     {
+        $requiredFields = $config->get('signup_required_fields');
```
[the remark this answers is not in the document]
Comment (xuwhite) [reply tests/Utils/SignUpConfig.php:14]: done

File: src/Factories/User.php (hunk 2 above)
Comment (xuwhite) [reply src/Factories/User.php:73]: done

File: tests/Utils/SignUpConfig.php (hunk 3 above)
Comment (xuwhite) [reply tests/Utils/SignUpConfig.php:42]: what same?

File: config/config.default.php (hunk 4 above)
Comment (xuwhite) [reply config/config.default.php:238]: done

**Comment by xuwhite** [comment]
> Another point is that
> 
>     * a user currently can remove some data after sign up in their profile
> 
>     * There is no hint implemented that notifies that some fields are missing when the config gets changed during an evnet

that is another feature pls make an issue for that

File: tests/Utils/SignUpConfig.php (hunk 3 above)
Comment (xuwhite) [reply tests/Utils/SignUpConfig.php:42]: done

**Comment by MyIgel** [comment]
> that is another feature pls make an issue for that

The notification: Maybe.
the fact that the settings have different requirements on the profile page not so much.

**Comment by xuwhite** [comment]
> > that is another feature pls make an issue for that
> 
> The notification: Maybe. the fact that the settings have different requirements on the profile page not so much -> That's just inconsistent

done

**Review by MyIgel — COMMENTED** [review]
@xuwhite  I am still getting a "plz configure your tshirt" message even when tshirt is not required, do we want that?
Screenshot_20231103_134938

And as we are on it: should we eventually implement that message for any required field?

**Pull request merged**